### PR TITLE
PR #42 (saveMetaBoxes--integration_tests)

### PR DESCRIPTION
**Pull request by rgadon107**
## PR Summary

`save_meta_boxes()` either deletes or updates custom field values in the database when called. It contains 5 dependent functions: 

	- get_meta_box_keys(),
	- get_meta_box_id(),
	- is_okay_to_save_meta_box(), 
	- configStore\getConfigParameter(), and 
	- save_custom_fields(). 

`save_custom_fields()` is a principal component of the meta box generator within the Central Hub plugin. The function (relative to it's Git repo) is located in:

>`wp-content/mu-plugins/central-hub/src/meta-data/meta-box.php`.

This PR contains the _integration_ tests for `save_meta_boxes()`. The _unit_ tests for the function were previously submitted in PR #33 (saveMetaBoxes) | unit tests only.

### CODE REVIEW

File: mu-plugins/central-hub/tests/phpunit/integration/meta-data/saveMetaBoxes.php [hunk 1]
```diff
@@ -0,0 +1,322 @@
...
+			loadConfig( $store_key, $config_to_store );
+		}
+		$_POST = [
+			'post_ID'            => $this->post,
+			'post_status'        => 'publish',
+			'members'            => [],
+			'members_nonce_name' => wp_create_nonce( 'members_nonce_action' ),
+		];
+
+		$this->assertNull( save_meta_boxes( $this->post ) );
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/tests/phpunit/integration/meta-data/saveMetaBoxes.php]: @rgadon107 This test currently doesn't test anything. Why? The only assertion is that the function returns `null`. When does the function have this behavior? Always. It always returns `null` whether it saves, deletes, or doesn't do anything. Therefore, this assertion doesn't tell us anything.

What do you want to test here?

The intent is in the method's name. You are testing that it does not save post meta _when_ there is no configuration.

Okay, so how would you test this?

1. Add custom fields into the `$_POST` array.
2. Run `save_meta_boxes()`.
3. Then test that the post meta you had in `$_POST` was not saved (hint use `get_post_meta()`.

File: mu-plugins/central-hub/tests/phpunit/integration/meta-data/saveMetaBoxes.php [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/tests/phpunit/integration/meta-data/saveMetaBoxes.php]: @rgadon107 Like above, this test doesn't validate anything.

In this one, you are attempting to test that it does not save (or update) when the configs exist but there's nothing to save in the `$_POST`.

Let's think about how to test it. We want to validate that it didn't save. There are a couple of ways to double check that:

1. None of the configured post meta exists in the db _before_ running `save_meta_boxes()`.
2. There are values in the db for the configured post meta; but they don't change after running `save_meta_boxes()`.

 Notice that either method tests that a save or update does not occur.

What do you think?

File: mu-plugins/central-hub/tests/phpunit/integration/meta-data/saveMetaBoxes.php [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+		$this->assertSame( '1', get_post_meta( $this->post, 'tour_number', true ) );
+
+		save_meta_boxes( $this->post );
+
+		// After function call, check that post meta keys are deleted from the database. 
+		$this->assertSame( [], $_POST['members'] );
+		$this->assertSame( '', get_post_meta( $this->post, 'role', true ) );
+		$this->assertSame( '', get_post_meta( $this->post, 'residence_city', true ) );
+		$this->assertSame( '', get_post_meta( $this->post, 'residence_state', true ) );
+		$this->assertSame( '', get_post_meta( $this->post, 'tour_number', true ) );
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/tests/phpunit/integration/meta-data/saveMetaBoxes.php]: @rgadon107 Great job here!

File: mu-plugins/central-hub/tests/phpunit/integration/meta-data/saveMetaBoxes.php [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+
+		// After function call, check that post meta keys are deleted from the database.
+		$this->assertSame( '', $_POST['members']['role'] );
+		$this->assertSame( '', $_POST['members']['residence_city'] );
+		$this->assertSame( '', $_POST['members']['residence_state'] );
+		$this->assertSame( '', $_POST['members']['tour_number'] );
+		$this->assertSame( '', get_post_meta( $this->post, 'role', true ) );
+		$this->assertSame( '', get_post_meta( $this->post, 'residence_city', true ) );
+		$this->assertSame( '', get_post_meta( $this->post, 'residence_state', true ) );
+		$this->assertSame( '', get_post_meta( $this->post, 'tour_number', true ) );
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/tests/phpunit/integration/meta-data/saveMetaBoxes.php]: @rgadon107 Great job here!

File: mu-plugins/central-hub/tests/phpunit/integration/meta-data/saveMetaBoxes.php [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+
+		save_meta_boxes( $this->post );
+
+		// After function call, check that database was updated with post meta.
+		$this->assertSame( 'Soprano', get_post_meta( $this->post, 'role', true ) );
+		$this->assertSame( 'Chicago', get_post_meta( $this->post, 'residence_city', true ) );
+		$this->assertSame( 'IL', get_post_meta( $this->post, 'residence_state', true ) );
+		$this->assertSame( '3', get_post_meta( $this->post, 'tour_number', true ) );
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/tests/phpunit/integration/meta-data/saveMetaBoxes.php]: @rgadon107 Great job here!

File: mu-plugins/central-hub/tests/phpunit/integration/meta-data/saveMetaBoxes.php [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+	}
+
+	/**
+	 * Test save_meta_boxes() should not save when no meta box keys are in the Config Store.
+	 */
+	public function test_should_not_save_when_no_meta_box_keys_are_in_store() {
+		$configs = [];
+		foreach ( $configs as $store_key => $config_to_store ) {
+			loadConfig( $store_key, $config_to_store );
+		}
```
[the remark this answers is not in the document]
Comment (rgadon107) [reply mu-plugins/central-hub/tests/phpunit/integration/meta-data/saveMetaBoxes.php]: Removed.

File: mu-plugins/central-hub/tests/phpunit/integration/meta-data/saveMetaBoxes.php (hunk 1 above)
Comment (rgadon107) [reply mu-plugins/central-hub/tests/phpunit/integration/meta-data/saveMetaBoxes.php]: First integration test method was updated, passed, and committed.

File: mu-plugins/central-hub/tests/phpunit/integration/meta-data/saveMetaBoxes.php (hunk 2 above)
Comment (rgadon107) [reply mu-plugins/central-hub/tests/phpunit/integration/meta-data/saveMetaBoxes.php]: I edited the second integration test to test _both_ use cases: 

1) None of the configured post meta exists in the db before running `save_meta_boxes()`. DB is not updated.
2) There are values in the DB for the configured post meta. Those values don't update after running `save_meta_boxes()`.

Integration test passed, and changed committed.

File: mu-plugins/central-hub/tests/phpunit/integration/meta-data/saveMetaBoxes.php (hunk 4 above)
Comment (rgadon107) [reply mu-plugins/central-hub/tests/phpunit/integration/meta-data/saveMetaBoxes.php]: Thank you!

File: mu-plugins/central-hub/tests/phpunit/integration/meta-data/saveMetaBoxes.php (hunk 5 above)
Comment (rgadon107) [reply mu-plugins/central-hub/tests/phpunit/integration/meta-data/saveMetaBoxes.php]: +1 ;-)

File: mu-plugins/central-hub/tests/phpunit/integration/meta-data/saveMetaBoxes.php (hunk 3 above)
Comment (rgadon107) [reply mu-plugins/central-hub/tests/phpunit/integration/meta-data/saveMetaBoxes.php]: +1 ;-)

File: mu-plugins/central-hub/tests/phpunit/integration/meta-data/saveMetaBoxes.php [hunk 7]
```diff
@@ -0,0 +1,374 @@
...
+			],
+			'members_nonce_name' => wp_create_nonce( 'members_nonce_action' ),
+		];
+		// Before function call, check that database is empty of post meta for this post.
+		$this->assertSame( '', get_post_meta( $this->post, 'role', true ) );
+		$this->assertSame( '', get_post_meta( $this->post, 'residence_city', true ) );
+		$this->assertSame( '', get_post_meta( $this->post, 'residence_state', true ) );
+		$this->assertSame( '', get_post_meta( $this->post, 'tour_number', true ) );
+
+		$this->assertNull( save_meta_boxes( $this->post ) );
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/tests/phpunit/integration/meta-data/saveMetaBoxes.php]: @rgadon107 I’d suggest removing the assertion here and just invoking the save function.

File: mu-plugins/central-hub/tests/phpunit/integration/meta-data/saveMetaBoxes.php [hunk 8]
```diff
@@ -0,0 +1,374 @@
...
+			'post_status'        => 'publish',
+			'members'            => [
+				'role'            => '',
+				'residence_city'  => '',
+				'residence_state' => '',
+				'tour_number'     => '',
+			],
+			'members_nonce_name' => wp_create_nonce( 'members_nonce_action' ),
+		];
+		$this->assertSame( 'Bass/Baritone', get_post_meta( $this->post, 'role', true ) );
```
Comment (hellofromtonya) [inline mu-plugins/central-hub/tests/phpunit/integration/meta-data/saveMetaBoxes.php]: @rgadon107 please add an empty line between the set up code and the assertions. Why? It helps with readability to see where things are set up vs. where the actual test assertions start.

Please make this change for all of the methods in this file. Thanks.

File: mu-plugins/central-hub/tests/phpunit/integration/meta-data/saveMetaBoxes.php (hunk 7 above)
Comment (rgadon107) [reply mu-plugins/central-hub/tests/phpunit/integration/meta-data/saveMetaBoxes.php]: Glad you provided a second set of eyes to look the file over. I missed that 2nd assertion. Done.

File: mu-plugins/central-hub/tests/phpunit/integration/meta-data/saveMetaBoxes.php (hunk 8 above)
Comment (rgadon107) [reply mu-plugins/central-hub/tests/phpunit/integration/meta-data/saveMetaBoxes.php]: Done.

**Pull request merged**